### PR TITLE
[Model] Add BAGEL step execution and continuous batching

### PR DESCRIPTION
**Pull request by Sky-Trigger**
## Purpose

This PR adds step execution and step-wise continuous batching for BAGEL image
generation while preserving BAGEL's existing complete-request execution path.

The implementation adds BAGEL's `prepare_encode()`, `denoise_step()`,
`step_scheduler()`, and `post_decode()` lifecycle and packs compatible requests
into one denoising call. Prompt/KV context, scheduler progress, CFG branches,
CFG normalization, and optional trajectory outputs remain request-local.
Packed token indexes are rebased for each request, and per-request KV lengths
are preserved when caches are merged.

Admission separates requests with incompatible effective image sizes or BAGEL
CFG/renormalization settings. For img2img, the effective output size is derived
from the input image before scheduler admission.

The existing non-step path remains available. Step execution in this PR covers
image generation only; BAGEL text output continues to use its existing
complete-request text-generation path.

## Deployment behavior

Image-generation step execution is supported by all three in-repo BAGEL deploy
configs. Their text and image paths behave as follows:

| Deploy config | Request/output path | Stages used | Step-execution behavior |
|---|---|---|---|
| `bagel.yaml` | Text output (`text2text` or `img2text`) | Stage 0 only | Stage 0 uses the existing autoregressive Thinker path to generate text. The diffusion stage is not used, so step execution does not participate. |
| `bagel.yaml` | Image output (`text2img` or `img2img`) | Stage 0 -> Stage 1 | Stage 0 performs the existing prompt/image processing and transfers the KV context after prefill. Stage 1 receives that context and performs image denoising with step execution. |
| `bagel_think.yaml` | Text output (`text2text` or `img2text`) | Stage 0 only | Stage 0 uses the existing autoregressive Thinker path to generate text. The diffusion stage is not used, so step execution does not participate. |
| `bagel_think.yaml` | Image output (`text2img` or `img2img`) | Stage 0 -> Stage 1 | Stage 0 keeps the existing think-token decoding and transfers KV after autoregressive decoding. Only Stage 1 image denoising uses step execution. |
| `bagel_single_stage.yaml` | Text output (`text2text` or `img2text`) | Stage 0 only | The request stays in the diffusion worker but is explicitly routed out of step execution. The existing complete-request `BagelPipeline.forward()` path calls `Bagel.generate_text()`, matching behavior without `--step-execution`. |
| `bagel_single_stage.yaml` | Image output (`text2img` or `img2img`) | Stage 0 only | The self-contained BAGEL diffusion stage performs preparation, step-wise denoising, scheduling, and decoding. Compatible requests can be continuously batched between denoising steps. |

`--step-execution` is applied only to diffusion stages and does not change the
autoregressive stages in the two-stage topologies. To enable continuous
batching, set `max_num_seqs` on Stage 1 for `bagel.yaml` or `bagel_think.yaml`,
and on Stage 0 for `bagel_single_stage.yaml`.

Current scope and constraints:

- BAGEL step execution supports the existing standard and CFG-parallel paths.
- Sequence parallelism and diffusion cache backends are not currently supported
  together with BAGEL step execution.
- Only requests with compatible image geometry and BAGEL CFG/renormalization
  settings are co-batched.

## Ascend benchmark

The latest benchmark measured the Stage 0 diffusion path from
`bagel_single_stage.yaml` under the following fixed setup:

- Model: `/mnt/sfs_turbo/BAGEL-7B-MoT`
- Devices: Ascend NPU `0,1,2,3`
- Tensor parallel size: 4
- Replicas: 1
- Requests: 64 compatible text-to-image requests
- Client concurrency: 32
- Resolution: 512 x 512
- Inference steps: 20
- Warmup: 32 requests at 2 inference steps
- Repeats: 3 per configuration

All measured runs completed `64/64` requests with zero failures. The table
reports the mean across the three runs:

| Configuration | `max_num_seqs` | Throughput | Duration | Mean latency | P95 latency | P99 latency | Peak memory |
|---|---:|---:|---:|---:|---:|---:|---:|
| Complete-request execution (`legacy_seq1`) | 1 | 0.341971 req/s | 187.173 s | 70.663 s | 94.088 s | 94.468 s | 10,908 MB |
| Step execution (`step_seq1`) | 1 | 0.369923 req/s | 173.013 s | 65.929 s | 87.515 s | 87.872 s | 10,912 MB |
| Step execution with continuous batching (`step_seq32`) | 32 | 0.696870 req/s | 91.846 s | 45.094 s | 46.935 s | 46.937 s | 18,800 MB |

With `max_num_seqs=1`, step execution improved throughput by **8.17%** over
complete-request execution (`0.341971 -> 0.369923 req/s`).

With `max_num_seqs=32`, continuous batching achieved:

- **2.0378x throughput**, or **+103.78%**, over complete-request execution.
- **+88.38% throughput** over step execution with `max_num_seqs=1`.
- **50.9% lower total duration** (`187.173 -> 91.846 s`).
- **36.2% lower mean latency** (`70.663 -> 45.094 s`).
- **50.1% lower P95 latency** (`94.088 -> 46.935 s`).

The higher batching capacity increases peak memory from approximately
10.9 GB to 18.8 GB. The result therefore demonstrates the throughput and
latency benefit of BAGEL step-wise continuous batching together with its
expected memory tradeoff.

### CODE REVIEW

**Comment by Sky-Trigger** [comment]
@SamitHuang  pls take a look

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 1]
```diff
@@ -2416,7 +2490,9 @@ def forward(
             extra_inputs["packed_vae_token_indexes"] = packed_vae_token_indexes
             extra_inputs["packed_text_indexes"] = packed_text_indexes
 
-        use_cfg = cfg_text_scale > 1.0
+        use_cfg = cfg_text_scale > 1.0 or (
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:2493]: If `use_cfg` evaluates to `True` because `cfg_text_scale > 1.0` or `cfg_text_scales` has values > 1.0, but `cfg_branch_pids` or `cfg_branch_caches` is `None`, the forward pass enters the single-forward `else` block and leaves `cfg_text_v_t` as `None`. When execution reaches `if use_cfg` at line 2542, it calls `_combine_cfg` with `cfg_text_v_t=None` and fails with a `TypeError`. Ensure `use_cfg` also requires `cfg_branch_pids is not None` and `cfg_branch_caches is not None` before attempting CFG combination.

**Review by SamitHuang — COMMENTED** [review]
## Review Summary

This PR adds BAGEL step execution and continuous batching support for image generation while preserving the single-request forward path.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 2]
```diff
@@ -875,6 +924,318 @@ def vae_transforms(img):
...
+        self,
+        state: StepRequestState,
+        **kwargs: object,
+    ) -> StepRequestState:
+        """Populate *state* with BAGEL inputs, latents, timesteps, and CFG config."""
+        del kwargs
+        sampling = state.sampling
+        prompt = state.prompt if state.prompt is not None else ""
+        if isinstance(prompt, dict):
+            if "text" in (prompt.get("modalities") or []):
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:937]: This check only inspects `prompt.get("modalities")` when `prompt` is a dictionary. In vLLM/OpenAI serving, `prompt` can be a string while text output modalities or sampling choices are specified in `state.sampling` or request parameters. Bypassing this check when `prompt` is a string allows unsupported text output requests to reach `_forward_single`, which attempts VAE latent generation and fails downstream with unexpected errors.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 3]
```diff
@@ -875,6 +924,318 @@ def vae_transforms(img):
...
+            "packed_seqlens": torch.cat(packed_seqlens),
+            "packed_position_ids": torch.cat([item["packed_position_ids"] for item in generation_inputs]),
+        }
+
+    def _build_denoise_kwargs(
+        self,
+        input_batch: InputBatch,
+        states: Sequence[StepRequestState],
+    ) -> dict[str, Any]:
+        if self.bagel._sp_size > 1:
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:1026]: The sequence parallelism check is placed inside `_build_denoise_kwargs`, which executes during `denoise_step` on every inference step. When a server is launched with `--sequence-parallel-size > 1`, `prepare_encode` succeeds and the request fails only when the first denoising step begins. Consider validating `_sp_size` in `prepare_encode` or during pipeline initialization so invalid sequence parallel requests fail early during state setup.

File: vllm_omni/diffusion/models/bagel/pipeline_bagel.py [hunk 4]
```diff
@@ -875,6 +924,318 @@ def vae_transforms(img):
...
+
+    def _denoise_step_cfg_parallel(self, denoise_kwargs: dict[str, Any]) -> torch.Tensor:
+        cfg_branch_pids = denoise_kwargs["cfg_branch_pids"]
+        cfg_branch_caches = denoise_kwargs["cfg_branch_caches"]
+        if cfg_branch_pids is None or cfg_branch_caches is None:
+            return self.bagel.forward(**denoise_kwargs)
+
+        cfg_world_size = get_classifier_free_guidance_world_size()
+        num_branches = len(cfg_branch_pids)
+        if cfg_world_size == 3 and num_branches == 2:
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/bagel/pipeline_bagel.py:1113]: Raising a `ValueError` when `cfg_world_size == 3` and `num_branches == 2` will crash inference when a server deployed with `--cfg-parallel-size 3` receives standard text-to-image requests (where `cfg_img_scale` is 1.0). Since `CFGParallelMixin` natively handles `cfg_world_size > n_branches` by running dummy forward passes on idle ranks during `all_gather`, this restriction unnecessarily prevents serving text-only CFG requests on 3-GPU CFG deployments. Consider allowing idle ranks to participate via `CFGParallelMixin` or falling back gracefully.

**Comment by Sky-Trigger** [comment]
updated， PTAL @SamitHuang @princepride

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 5]
```diff
@@ -995,6 +1027,18 @@ def _execute_stepwise(
                     request_id=new_req.req.request_id,
                     metadata=new_req.diffusion_kv_metadata,
                 )
+        non_step_requests = [
```
Comment (princepride) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py:1137]: The non-step fallback returns before the normal finished-request cleanup. If a running step request is aborted and the next scheduler wave admits a full-forward text request, finished_req_ids is never applied to state_cache or paged diffusion KV cleanup, and the previous input_batch may continue retaining its tensors. Please move the common finished-request cleanup before the non-step dispatch and invalidate stale step-batch state when switching to a pure full-forward wave. A step-request → abort/finish → text-fallback lifecycle test would cover this.

File: vllm_omni/diffusion/diffusion_engine.py [hunk 6]
```diff
@@ -1141,6 +1141,10 @@ def _dummy_run(self):
             height=512,
             width=512,
             guidance_scale=0.0,
+            # Step execution must exercise at least one denoising iteration.
+            # Some pipelines (for example BAGEL) perform
+            # ``num_inference_steps - 1`` scheduler updates.
+            num_inference_steps=2 if self.step_execution else 1,
```
Comment (princepride) [inline vllm_omni/diffusion/diffusion_engine.py]: The num_inference_steps - 1 behavior is a bit surprising and confusing,😂 can you help me change the Bagel's denoise step calculate.

**Review by princepride — CHANGES_REQUESTED** [review]
PTAL

File: vllm_omni/diffusion/worker/diffusion_model_runner.py (hunk 5 above)
Comment (Sky-Trigger) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py:1137]: using self._cleanup_finished_step_requests(scheduler_output) for now.

File: vllm_omni/diffusion/diffusion_engine.py (hunk 6 above)
Comment (Sky-Trigger) [reply vllm_omni/diffusion/diffusion_engine.py]: updated,  PTAL. @princepride

**Review by princepride — APPROVED** [review]
LGTM

**Comment by Sky-Trigger** [comment]
ready for merge please @princepride

**Comment by david6666666** [comment]
please fix merge CI https://buildkite.com/vllm/vllm-omni/builds/14611/canvas?sid=01a06a12-bf0c-469c-a4cf-056454da21f6&tab=output

**Comment by david6666666** [comment]
@Sky-Trigger

**Comment by Sky-Trigger** [comment]
https://github.com/vllm-project/vllm-omni/issues/7031 @david6666666  @yenuo26  @princepride   It looks like the current merge CI conflicts with our changes to the actual number of denoising steps in BAGEL. I'll open a new PR to revert it to the original denoising step counting method.

**Comment by Sky-Trigger** [comment]
https://github.com/vllm-project/vllm-omni/pull/7049 @david6666666  @yenuo26  PTAL
